tests/unattended/unit/framework/bach.sh

```diff
@@ -0,0 +1,642 @@
+} >&2
+export -f @die
+
+if [[ -z "${BASH_VERSION:-}" ]] || [[ "${BASH_VERSINFO[0]}" -lt 4 ]]; then
+    @die "Bach Testing Framework only support Bash v4+!"
+fi
+
+if [[ "${BACH_DEBUG:-}" != true ]]; then
+    function @debug() {
+        :
+    }
+else
+    exec 8>&2
+    function @debug() {
+        builtin printf '[DEBUG] %s\n' "$*"
+    } >&8
+fi
+export -f @debug
+
+function bach-real-path() {
+    declare folder name="$1"
+    declare altname="${name#*|}"
+    name="${name%|*}"
+    for folder in "${bach_origin_paths[@]}"; do
+        if [[ -x "$folder/$name" ]]; then
+           builtin echo "$folder/$name"
+           return 0
+        elif [[ "$name" != "$altname" && -x "$folder/$altname" ]]; then
+            builtin echo "$folder/$altname"
+            return 0
+        fi
+    done
+    return 1
+}
+export -f bach-real-path
+
+export BACH_DEV_STDIN=""
+
+function bach_restore_stdin() {
+    if [[ ! -t 0 ]]; then
+        declare name
+        [[ -n "$BACH_DEV_STDIN" ]] || for name in /dev/ptmx /dev/pts/ptmx /dev/ttyv[0-9a-f]; do
+            if [[ -r "$name" && -c "$name" ]]; then
+                ls -l "$name" >&2
+                BACH_DEV_STDIN="$name"
+                break
+            fi
+        done
+        exec 0<&-
+        exec 0<"$BACH_DEV_STDIN"
+    fi
+}
+
+function bach_initialize(){
+    enable -n alias bg bind dirs disown fc fg hash help history jobs kill suspend times ulimit umask unalias wait
+
+    declare util name util_path
+
+    declare -a bash_builtin_cmds=(cd echo enable popd pushd pwd shopt test trap type)
+
+    for name in . command exec false set true unset "${bash_builtin_cmds[@]}"; do
+        eval "function @${name}() { builtin $name \"\$@\"; } 8>/dev/null; export -f @${name}"
+    done
+
+    for name in eval; do
+        eval "function @${name}() { builtin $name \"\$@\"; }; export -f @${name}"
+    done
+
+    function @source() {
+        declare script="$1"
+        shift
+        builtin source "$script" "$@"
+    }
+
+    declare -a bach_core_utils=(cat chmod cut diff find env grep ls "shasum|sha1sum" mkdir mktemp rm rmdir sed sort tee touch which xargs)
+
+    for util in "${bach_core_utils[@]}"; do
+        if [[ "$util" == "shasum|"* && "$BACH_OS_NAME" == FreeBSD ]]; then
+            util="shasum|sha1"
+        fi
+        name="${util%|*}"
+        util_path="$(bach-real-path "$util")"
+        eval "[[ -n \"${util_path}\" ]] || @die \"Fatal, CAN NOT find '$name' in \\\$PATH\"; function @${name}() { \"${util_path}\" \"\$@\"; } 8>/dev/null; export -f @${name}"
+    done
+
+    bach_restore_stdin
+    @mockall "${bash_builtin_cmds[@]}" source .
+
+    eval "$(builtin export | while read -rs name; do
+        name="${name%%=*}"
+        name="${name##* }"
+        [[ "${name^^}" != BACH_* ]] || continue
+        builtin echo "unset '$name' || builtin true"
+    done)"
+    builtin export LANG=C TERM=vt100
+    unset name util_path
+}
+
+function @real() {
+    declare name="$1" real_cmd
```

Comment:
Same

Reply:
`bach.sh` is the external framework. We will not modify it.